### ISSUE TITLE
`GridPatch` with `pin_memory=True` significantly slow-down in following epochs

### ISSUE THREAD

**Issue by drbeh** [Project-MONAI/MONAI#6082]
Reported by a user (@kenza-bouzid) [here](https://github.com/Project-MONAI/MONAI/pull/6055#discussion_r1116970094): 

When using num_workers>1 and pin_memory=True, training time increases exponentially over epochs 
<img width="423" alt="image" src="https://user-images.githubusercontent.com/37396332/221173718-6639f259-f2a2-4ece-9094-369980853e98.png">
<img width="429" alt="image" src="https://user-images.githubusercontent.com/37396332/221173749-4635aebc-c33a-4fa2-bc84-cd7d14c8ec9a.png">
I had profiled/timed all intermediate steps, and found out that GridPatch was the guilty one 
<img width="466" alt="image" src="https://user-images.githubusercontent.com/37396332/221174364-580c436b-5d66-4315-812b-1beb1665f21d.png">
So I timed all intermediate step in the tranform 
 
<img width="621" alt="image" src="https://user-images.githubusercontent.com/37396332/221181608-7913298b-6bca-40df-96fd-b2137b8fe670.png">
<img width="675" alt="image" src="https://user-images.githubusercontent.com/37396332/221181800-e5ca3d07-f821-4944-a17b-121e13168e98.png">

 
<img width="689" alt="image" src="https://user-images.githubusercontent.com/37396332/221182074-4d9a3241-55e5-4be7-80e2-313dd81e36e3.png">

It turned out that the operation that was taking too long was a memory allocation by np.array
https://github.com/Project-MONAI/MONAI/blob/a2ec3752f54bfc3b40e7952234fbeb5452ed63e3/monai/transforms/spatial/array.py#L3279
I eventually fixed it by setting pin_memory=False, 
which I explain by cuda memory allocation being more expensive as it has to be allocated from the pinned memory

Note that I am dealing with particularly large slides ~50kx50k 

Any thoughts on this?

**Comment by drbeh** [comment]
@kenza-bouzid just to double confirm that when you set `pin_memory=False` this issue goes away and it takes the same time (few seconds) in all epochs, right?

**Comment by kenza-bouzid** [comment]
Correct! I don't have a proper explanation why pin_memory=False fixes this issue tbh.

**Comment by drbeh** [comment]
@ericspod @Nic-Ma @wyli @rijobro do you have any insights here on why setting `pin_memory=True` in dataloader causes this behavior?

It seems that these conversion to and from numpy arrays have some effects here but I thought pinning memory is being done in the last step of dataloader so I don't see why it has anything to do with it.
Also there have been some issues with pin memory in pytorch in the past, like [this one](https://github.com/pytorch/pytorch/issues/25010) from years ago, so can it be a pytorch issue?

**Comment by ericspod** [comment]
I honestly am not sure what the issue here would be. I recall there being other issues we've encountered in the past with pinned memory so I typically avoid it. I can only speculate it could be the pinned segment's lifetime continues after the `np.array` conversion because of Pytorch's buffer reuse or some other issue in Pytorch or CUDA itself.
